packages/angular/ssr/src/utils/inline-critical-css.ts

```diff
@@ -0,0 +1,214 @@
+  '    target.removeAttribute(CSP_MEDIA_ATTR);',
+
+  // Remove onload listener when there are no longer styles that need to be loaded.
+  '    if (!document.head.querySelector(`link[${CSP_MEDIA_ATTR}]`)) {',
+  `      documentElement.removeEventListener('load', listener);`,
+  '    }',
+  '  };',
+
+  //  We use an event with capturing (the true parameter) because load events don't bubble.
+  `  documentElement.addEventListener('load', listener, true);`,
+  '})();',
+].join('\n');
+
+/** Partial representation of an `HTMLElement`. */
+interface PartialHTMLElement {
+  getAttribute(name: string): string | null;
+  setAttribute(name: string, value: string): void;
+  hasAttribute(name: string): boolean;
+  removeAttribute(name: string): void;
+  appendChild(child: PartialHTMLElement): void;
+  insertBefore(newNode: PartialHTMLElement, referenceNode?: PartialHTMLElement): void;
+  remove(): void;
+  name: string;
+  textContent: string;
+  tagName: string | null;
+  children: PartialHTMLElement[];
+  next: PartialHTMLElement | null;
+  prev: PartialHTMLElement | null;
+}
+
+/** Partial representation of an HTML `Document`. */
+interface PartialDocument {
+  head: PartialHTMLElement;
+  createElement(tagName: string): PartialHTMLElement;
+  querySelector(selector: string): PartialHTMLElement | null;
+}
+
+/** Signature of the `Critters.embedLinkedStylesheet` method. */
+type EmbedLinkedStylesheetFn = (
+  link: PartialHTMLElement,
+  document: PartialDocument,
+) => Promise<unknown>;
+
+export class InlineCriticalCssProcessor extends Critters {
+  private initialEmbedLinkedStylesheet: EmbedLinkedStylesheetFn;
+  private addedCspScriptsDocuments = new WeakSet<PartialDocument>();
+  private documentNonces = new WeakMap<PartialDocument, string | null>();
+
+  // Inherited from `Critters`, but not exposed in the typings.
+  protected declare embedLinkedStylesheet: EmbedLinkedStylesheetFn;
+
+  constructor(readonly outputPath?: string) {
+    super({
+      logger: {
+        // eslint-disable-next-line no-console
+        warn: (s: string) => console.warn(s),
+        // eslint-disable-next-line no-console
+        error: (s: string) => console.error(s),
+        info: () => {},
+      },
+      logLevel: 'warn',
+      path: outputPath,
+      publicPath: undefined,
+      compress: false,
+      pruneSource: false,
+      reduceInlineStyles: false,
+      mergeStylesheets: false,
+      // Note: if `preload` changes to anything other than `media`, the logic in
+      // `embedLinkedStylesheetOverride` will have to be updated.
+      preload: 'media',
+      noscriptFallback: true,
+      inlineFonts: true,
+    });
+
+    // We can't use inheritance to override `embedLinkedStylesheet`, because it's not declared in
+    // the `Critters` .d.ts which means that we can't call the `super` implementation. TS doesn't
+    // allow for `super` to be cast to a different type.
+    this.initialEmbedLinkedStylesheet = this.embedLinkedStylesheet;
```

Comment:
Updated, although I had to do it in another way since Critters is exported as a default class, and type merging isn't supported in that scenario.

Reply:
Ah, that's unfortunate, but extending a custom base class isn't too bad. It would be helpful to still include a comment explaining why we need this declaration merge.